File: backend/claire/projects/concordance.py
"""Concordance humain ↔ LLM et LLM ↔ LLM (feature « point 4 »).

Agrège, pour un annotateur donné et sur l'ensemble de ses annotations d'une
campagne, l'accord entre SON annotation et chaque modèle LLM pré-annoté, ainsi
qu'entre les modèles eux-mêmes.

Modèles (cohérents avec `iaa.py` et le frontend) :
  - HUMAIN : per-sentence EXACT — la phrase *i* porte le thème de la clause ancrée
    exactement en *i*, sinon ``None`` (pas de forward-fill).
  - LLM : modèle de bloc — le thème forward-fill depuis l'ancre du juge.

Mesure = accord sur l'INTERSECTION : parmi les phrases co-couvertes par les deux
parties, la fraction qui porte le même thème. Le support (``n``) est explicite.
"""

from __future__ import annotations

from collections import defaultdict

from claire.imports.models import PreAnnotation

from .iaa import _theme_vector


def _judge_vector(preclauses, n_sentences: int) -> list[str | None]:
    """Vecteur par phrase d'un juge (forward-fill, modèle de bloc)."""
    items = sorted((c.anchor_index, c.theme_code) for c in preclauses)
    vec: list[str | None] = []
    cur: str | None = None
    k = 0
    for i in range(n_sentences):
        while k < len(items) and items[k][0] <= i:
            cur = items[k][1]
            k += 1
        vec.append(cur)
    return vec


def _intersection(a: list, b: list) -> tuple[int, int]:
    """(co-couvertes, accords) sur l'intersection de deux vecteurs alignés."""
    n = 0
    matches = 0
    for x, y in zip(a, b):
        if x is not None and y is not None:
            n += 1
            if x == y:
                matches += 1
    return n, matches


def _judge_vectors_for_document(project, document, n_sentences: int) -> dict[str, list]:
    """Vecteur forward-fill par juge pour un document (dernière version importée)."""
    pres = list(
        PreAnnotation.objects.filter(project=project, document=document)
        .order_by("imported_at")
        .prefetch_related("preclauses")
    )
    # Une seule entrée par juge : la PLUS RÉCENTE (order_by croissant → la dernière gagne).
    by_judge: dict[str, PreAnnotation] = {}
    for p in pres:
        by_judge[p.judge] = p
    return {
        judge: _judge_vector(p.preclauses.all(), n_sentences)
        for judge, p in by_judge.items()
    }


def project_concordance(project, user) -> dict:
    """Agrège la concordance de l'annotateur ``user`` sur ses annotations du projet.

    Renvoie ``per_judge`` (accord humain↔juge, trié décroissant), ``best_match``
    (juge le plus concordant), ``llm_pairs`` + ``llm_mean_pct`` (LLM↔LLM), et le
    nombre de documents comparés. ``None`` si rien à comparer (ni annotation, ni juge).
    """
    annotations = list(
        project.annotations.filter(annotator=user).select_related("document")
    )

    # Accumulateurs humain↔juge et juge↔juge (sur l'ensemble des documents).
    hj_n: dict[str, int] = defaultdict(int)
    hj_m: dict[str, int] = defaultdict(int)
    jj_n: dict[tuple[str, str], int] = defaultdict(int)
    jj_m: dict[tuple[str, str], int] = defaultdict(int)
    documents_compared = 0
    human_covered = 0

    for ann in annotations:
        doc = ann.document
        n = doc.n_sentences
        if not n:
            continue
        judge_vecs = _judge_vectors_for_document(project, doc, n)
        if not judge_vecs:
            continue
        human = _theme_vector(ann, n)
        human_covered += sum(1 for t in human if t is not None)
        documents_compared += 1

        for judge, vec in judge_vecs.items():
            n_co, m = _intersection(human, vec)
            hj_n[judge] += n_co
            hj_m[judge] += m

        judges = sorted(judge_vecs)
        for i in range(len(judges)):
            for j in range(i + 1, len(judges)):
                a, b = judges[i], judges[j]
                n_co, m = _intersection(judge_vecs[a], judge_vecs[b])
                jj_n[(a, b)] += n_co
                jj_m[(a, b)] += m

    if documents_compared == 0:
        return None

    per_judge = []
    for judge in hj_n:
        n_co = hj_n[judge]
        pct = round((hj_m[judge] / n_co) * 100, 1) if n_co else None
        per_judge.append({"judge": judge, "pct": pct, "n": n_co, "matches": hj_m[judge]})
    # Tri : meilleurs accords d'abord, juges sans support (pct None) en fin.
    per_judge.sort(key=lambda p: (p["pct"] is None, -(p["pct"] or 0)))

    with_support = [p for p in per_judge if p["pct"] is not None]
    best_match = (
        {"judge": with_support[0]["judge"], "pct": with_support[0]["pct"]}
        if with_support
        else None
    )

    llm_pairs = []
    for (a, b), n_co in jj_n.items():
        pct = round((jj_m[(a, b)] / n_co) * 100, 1) if n_co else None
        llm_pairs.append({"a": a, "b": b, "pct": pct, "n": n_co})
    pcts = [p["pct"] for p in llm_pairs if p["pct"] is not None]
    llm_mean_pct = round(sum(pcts) / len(pcts), 1) if pcts else None

    return {
        "per_judge": per_judge,
        "best_match": best_match,
        "llm_pairs": llm_pairs,
        "llm_mean_pct": llm_mean_pct,
        "documents_compared": documents_compared,
        "human_covered": human_covered,
    }

File: frontend/src/lib/concordance.ts
/**
 * Concordance (point 4) — calcul PUR (sans React) des KPIs d'accord entre l'annotateur
 * humain et chaque juge LLM, ainsi qu'entre les LLM eux-mêmes. Brique testée isolément.
 *
 * Modèles (cohérents avec le backend IAA et le triage) :
 *  - HUMAIN : per-sentence EXACT (C4) — la phrase *i* porte le thème de la clause ancrée
 *    EXACTEMENT en *i*, sinon « non couverte ». Pas de forward-fill (une phrase non
 *    annotée est réellement vide).
 *  - LLM : modèle de bloc — le thème forward-fill depuis l'ancre (réutilise `themeByIndex`).
 *
 * Mesure = accord sur l'INTERSECTION : parmi les phrases co-couvertes par les DEUX
 * parties (toutes deux non vides), la fraction qui porte le même thème. C'est la lecture
 * la plus pertinente « à quel point suis-je d'accord avec ce modèle » : tant que je n'ai
 * annoté que quelques phrases, on ne me pénalise pas pour celles que je n'ai pas encore
 * traitées (contrairement à un accord sur l'union). `n` (co-couvertes) rend le support
 * explicite.
 */

import { themeByIndex } from "./llmAgreement";

/** Vecteur EXACT (modèle humain C4) : thème à l'index exact, sinon null. Défensif. */
export function themeVectorExact(
  anchors: { anchorIndex: number; theme: string }[],
  n: number,
): (string | null)[] {
  if (!Number.isInteger(n) || n <= 0) return [];
  const out: (string | null)[] = new Array(n).fill(null);
  for (const a of anchors ?? []) {
    if (Number.isInteger(a.anchorIndex) && a.anchorIndex >= 0 && a.anchorIndex < n) {
      out[a.anchorIndex] = a.theme;
    }
  }
  return out;
}

/** Vecteurs par juge (forward-fill, modèle de bloc) depuis les pré-annotations. */
export function judgeVectorsFromPre(
  preByJudge: Record<string, { clauses?: { anchorIndex: number; themeCode: string }[] }>,
  n: number,
): Record<string, (string | null)[]> {
  const out: Record<string, (string | null)[]> = {};
  for (const [id, p] of Object.entries(preByJudge ?? {})) {
    out[id] = themeByIndex(
      (p?.clauses ?? []).map((c) => ({ anchorIndex: c.anchorIndex, theme: c.themeCode })),
      n,
    );
  }
  return out;
}

export interface PairAgreement {
  /** % d'accord sur l'intersection (0–100), ou null si aucune phrase co-couverte. */
  pct: number | null;
  /** Nombre de phrases co-couvertes (support de la mesure). */
  n: number;
  /** Nombre de phrases co-couvertes en accord. */
  matches: number;
}

/** Accord sur l'INTERSECTION de deux projections par phrase (mêmes indices). */
export function intersectionAgreement(
  a: (string | null)[],
  b: (string | null)[],
): PairAgreement {
  const len = Math.min(a.length, b.length);
  let n = 0;
  let matches = 0;
  for (let i = 0; i < len; i += 1) {
    if (a[i] != null && b[i] != null) {
      n += 1;
      if (a[i] === b[i]) matches += 1;
    }
  }
  return { n, matches, pct: n === 0 ? null : (matches / n) * 100 };
}

export interface JudgeConcordance {
  judge: string;
  pct: number | null;
  n: number;
  matches: number;
}

export interface LlmPair {
  a: string;
  b: string;
  pct: number | null;
  n: number;
}

export interface ConcordanceReport {
  /** Accord humain ↔ chaque juge, trié par % décroissant (juges sans support en fin). */
  perJudge: JudgeConcordance[];
  /** Juge le plus concordant avec l'humain (support > 0), ou null. */
  bestMatch: { judge: string; pct: number } | null;
  /** Accord LLM ↔ LLM, paire à paire. */
  llmPairs: LlmPair[];
  /** Moyenne des accords LLM ↔ LLM (paires avec support), ou null. */
  llmMeanPct: number | null;
  /** Nombre de phrases que l'humain a annotées (couverture humaine). */
  humanCovered: number;
}

/**
 * Rapport de concordance complet à partir du vecteur humain (exact) et des vecteurs
 * des juges (forward-fill). PUR et déterministe.
 */
export function concordanceReport(
  humanVec: (string | null)[],
  judgeVecs: Record<string, (string | null)[]>,
): ConcordanceReport {
  const judges = Object.keys(judgeVecs);

  const perJudge: JudgeConcordance[] = judges.map((j) => {
    const r = intersectionAgreement(humanVec, judgeVecs[j]!);
    return { judge: j, pct: r.pct, n: r.n, matches: r.matches };
  });
  // Tri : meilleurs accords d'abord ; les juges sans support (pct null) en dernier.
  perJudge.sort((x, y) => {
    if (x.pct == null && y.pct == null) return 0;
    if (x.pct == null) return 1;
    if (y.pct == null) return -1;
    return y.pct - x.pct;
  });

  const withSupport = perJudge.filter((p) => p.pct != null);
  const bestMatch =
    withSupport.length > 0
      ? { judge: withSupport[0]!.judge, pct: withSupport[0]!.pct as number }
      : null;

  const llmPairs: LlmPair[] = [];
  for (let i = 0; i < judges.length; i += 1) {
    for (let j = i + 1; j < judges.length; j += 1) {
      const a = judges[i]!;
      const b = judges[j]!;
      const r = intersectionAgreement(judgeVecs[a]!, judgeVecs[b]!);
      llmPairs.push({ a, b, pct: r.pct, n: r.n });
    }
  }
  const pcts = llmPairs.map((p) => p.pct).filter((p): p is number => p != null);
  const llmMeanPct =
    pcts.length === 0 ? null : pcts.reduce((s, v) => s + v, 0) / pcts.length;

  const humanCovered = humanVec.filter((t) => t != null).length;

  return { perJudge, bestMatch, llmPairs, llmMeanPct, humanCovered };
}
